src/controllers/authenticationController.ts

```diff
@@ -28,67 +28,117 @@ export default class AuthenticationController {
+    const errorOrBool = await this.sessionStorage.set(
+      user.token,
+      user,
+      timestamp
     );
+
+    if (isLeft(errorOrBool)) {
+      const error = errorOrBool.value;
+      return left(error);
+    }
+    const response = errorOrBool.value;
+
+    if (!response) {
+      return left(new Error("Error creating the user session"));
+    }
+    const url =
+      process.env.CLIENT_REDIRECTION_URL || "/profile.html?token={token}";
+    const urlWithToken = url.replace("{token}", user.token);
+
+    return right(urlWithToken);
   }
 
   /**
    * Retrieves the logout url from the IDP.
    */
-  public logout(req: express.Request, res: express.Response): void {
+  public async logout(req: express.Request): Promise<Either<Error, string>> {
     const errorOrUser = extractUserFromRequest(req);
 
-    errorOrUser.fold(
-      (error: Error) => this.return500Error(error, res),
-      (user: User) => {
-        // Delete the Redis token.
-        this.sessionStorage.del(user.token);
-
-        // Logout from SPID.
-        req.query = {};
-        req.query.entityID = user.spid_idp;
-
-        this.spidStrategy.logout(req, (err, request: express.Request) => {
-          if (!err) {
-            res.status(200).json({
-              logoutUrl: request
-            });
-          } else {
-            res.status(500).json({
-              message: err.toString()
-            });
-          }
-        });
+    if (isLeft(errorOrUser)) {
+      const error = errorOrUser.value;
+      return left(error);
+    }
+
+    const user = errorOrUser.value;
+
+    const errorOrBool = await this.sessionStorage.del(user.token);
+
+    if (isLeft(errorOrBool)) {
+      const error = errorOrBool.value;
+      return left(error);
+    }
+
+    const response = errorOrBool.value;
+
+    if (!response) {
+      return left(new Error("Error creating the user session"));
+    }
+
+    // Logout from SPID.
+    req.query = {};
+    req.query.entityID = user.spid_idp;
+
+    return await this.spidLogout(req);
+  }
+
+  /**
+   * Returns data about the current user session.
+   */
+  public async getSessionState(
+    req: express.Request
+  ): Promise<Either<Error, string>> {
+    const maybeToken = await this.extractTokenFromRequest(req);
+    if (isNone(maybeToken)) {
+      return left(new Error("No token in the request"));
+    }
+
+    const token = maybeToken.value;
+    const errorOrSession = await this.sessionStorage.get(token);
+    if (isLeft(errorOrSession)) {
+      // Previous token not found or expired, try to refresh.
+      const errorOrSessionState = await this.refreshUserToken(token);
+
+      if (isLeft(errorOrSessionState)) {
+        // Unable to refresh token or session not found.
+        const error = errorOrSessionState.value;
+        return left(error);
       }
-    );
+
+      // Return the new session information.
+      const sessionState = errorOrSessionState.value;
+      return right(JSON.stringify(sessionState));
```

Comment:
what's the reason for losing type information at this point? this function should return something like `Promise<Either<Error, Session>>` where `Sessions` is the super type of `sessionState` and `publicSession` (`ISessionState` ?)